### PR TITLE
Add typings to the cap-level-controller

### PR DESCRIPTION
**Pull request by jnatalzia**
### This PR will...
Add typescripting to the cap-level-controller.

### Why is this Pull Request needed?
Documentation as code as documentation

### Are there any points in the code the reviewer needs to double check?
All of it, I'm new

### Resolves issues:
?

### Checklist

~- [ ] changes have been done against master branch, and PR does not conflict~
- [x] changes have been done against LHLS branch and PR does not conflict
- [x] new unit / functional tests have been added (whenever applicable)
- [x] API or design changes are documented in API.md

### CODE REVIEW

File: src/controller/cap-level-controller.ts [hunk 1]
```diff
@@ -16,11 +25,11 @@ class CapLevelController extends EventHandler {
       Event.MEDIA_DETACHING);
 
     this.autoLevelCapping = Number.POSITIVE_INFINITY;
-    this.firstLevel = null;
     this.levels = [];
+    this.firstLevel = null;
     this.media = null;
     this.restrictedLevels = [];
-    this.timer = null;
+    this.timer = undefined;
```
Comment (jnatalzia) [inline src/controller/cap-level-controller.ts]: typescript expects number/undefined for interval

File: src/controller/cap-level-controller.ts [hunk 2]
```diff
@@ -30,18 +39,40 @@ class CapLevelController extends EventHandler {
...
-
-  onManifestParsed (data) {
+  onMediaAttaching (data: {
+    media: HTMLVideoElement
+  }) {
+    this.media = data.media instanceof HTMLVideoElement ? data.media : null;
+  }
+
+  onManifestParsed (data: {
+    altAudio: boolean,
```
Comment (jnatalzia) [inline src/controller/cap-level-controller.ts]: should i pull these event objects out of this file and create types in the `types` folder?

File: src/controller/cap-level-controller.ts [hunk 3]
```diff
@@ -30,18 +39,40 @@ class CapLevelController extends EventHandler {
...
+    audio: boolean,
+    audioTracks: Array<AudioTrack>,
+    firstLevel: number,
+    levels: Array<Level>,
+    stats: {
```
Comment (jnatalzia) [inline src/controller/cap-level-controller.ts]: is there already a type for this?

File: src/controller/cap-level-controller.ts (hunk 2 above)
Comment (johnBartos) [reply src/controller/cap-level-controller.ts]: Yeah that'd be great

File: src/controller/cap-level-controller.ts (hunk 3 above)
Comment (johnBartos) [reply src/controller/cap-level-controller.ts]: Yeah, `LoaderStats`. Just merged the stats overhaul PR so you'll have to rebase against that

File: src/controller/cap-level-controller.ts [hunk 4]
```diff
@@ -107,9 +142,9 @@ class CapLevelController extends EventHandler {
       return;
     }
     this.autoLevelCapping = Number.POSITIVE_INFINITY;
-    this.hls.firstLevel = this.getMaxLevel(this.firstLevel);
+    this.hls.firstLevel = this.getMaxLevel(this.firstLevel || -1);
```
Comment (johnBartos) [inline src/controller/cap-level-controller.ts]: Would it be possible to make `firstLevel` always a number (not null)?

Comment (jnatalzia) [reply src/controller/cap-level-controller.ts]: can do

**Comment by jnatalzia** [comment]
@johnBartos updated

**Pull request merged**